### PR TITLE
[SPARK-37769][SQL][FOLLOWUP] Filtering files if metadata columns are present in the data filter

### PR DESCRIPTION
**Pull request by Yaohua628**
### What changes were proposed in this pull request?
Follow-up PR of #34575. Filtering files if metadata columns are present in the data filter.

### Why are the changes needed?
Performance improvements.

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Existing UTs and a new UT.

### CODE REVIEW

**Comment by Yaohua628** [comment]
Hi @cloud-fan, here's metadata filtering PR, please take a look whenever you have a chance. Also, got a question: what about `df.inputFiles`? Do you have any idea how to make it work as well? Thanks!

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/FileMetadataStructSuite.scala [hunk 1]
```diff
@@ -288,6 +288,20 @@ class FileMetadataStructSuite extends QueryTest with SharedSparkSession {
     )
   }
 
+  metadataColumnsTest("filter on metadata and user data", schema) { (df, _, f1) =>
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/FileMetadataStructSuite.scala:291]: Does this test fail before?

Comment (cloud-fan) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/FileMetadataStructSuite.scala:291]: I think we should also check the physical plan and make sure the final file list is pruned.

Comment (Yaohua628) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/FileMetadataStructSuite.scala:304]: no, it won't fail before. I added this test just wanna make sure we only get metadata filters in `listFiles(...)` and do the correct filtering [here](https://github.com/apache/spark/pull/35055/files#diff-0d5b92aea54d993bd94249fdc9ba7d2616a3f4f04289ffaa2deb55a06585d4f9R78)

thanks for the suggestion! updated tests to check files!

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PartitioningAwareFileIndex.scala [hunk 2]
```diff
@@ -71,8 +73,34 @@ abstract class PartitioningAwareFileIndex(
...
+    val boundedFilterOpt = fileMetadataFilterOpt.map { fileMetadataFilter =>
+      Predicate.createInterpreted(fileMetadataFilter.transform {
+        case _: AttributeReference => BoundReference(0, METADATA_STRUCT, nullable = true)
+      })
+    }
+
+    def matchFileMetadataPredicate(f: FileStatus): Boolean = {
+      // use option.forall, so if there is no filter, return true
+      boundedFilterOpt.forall(_.eval(
+        InternalRow.fromSeq(Seq(InternalRow.fromSeq(Seq(
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PartitioningAwareFileIndex.scala:93]: Can we have a util method to create InternalRow from `Path`, `length: Long` and `modificationTime: Long`? To share code with `FileScanRDD`

**Review by cloud-fan — APPROVED** [review]

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PartitioningAwareFileIndex.scala (hunk 2 above)
Comment (Yaohua628) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PartitioningAwareFileIndex.scala:93]: could we have this util method and share the common code maybe along with the metadata schema pruning PR?

I am thinking after pruning, we don't need to always create `InternalRow` with all fields in both `PartitioningAwareFileIndex.listFiles(...)` and `FileScanRDD`. WDYT?

Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PartitioningAwareFileIndex.scala:93]: SGTM

**Comment by Yaohua628** [comment]
@cloud-fan Hi Wenchen, updated this PR on top of the schema pruning. 

Also, add 2 utility methods and tried to share more code between `FileScanRDD` and `PartitioningAwareFileIndex` to create/update metadata internal row. Please let me know what do you think, thanks!!

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
LGTM if tests pass

**Comment by Yaohua628** [comment]
> LGTM if tests pass

Thanks, passed!

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Comment by Yaohua628** [comment]
thanks, Wenchen!

**Pull request closed without merging**